### PR TITLE
Apply Shellcheck improvement suggestions

### PR DESCRIPTION
**Pull request by Leonidas-from-XIV**
I ran shellcheck on the configure script which suggested a lot of nice improvements, mostly about quoting and making sure spaces don't break the commands.

Other interesting improvements include removing variables that are never set (thus always empty, which made it work), so removing them makes it easier to see what is happening.

Edit: I also applied the `$?` suggestion since it makes the code a bit less fragile and fixed the POSIX compatibility, so it should be as portable as it was before.

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -676,32 +677,33 @@ for lib in $generated_META $lbytes; do
...
 
-for part in `cd src; echo *`; do
-    if [ -f "src/$part/META.in" ]; then
-	sed -e "s/@VERSION@/$version/g" \
-            -e "s/@REQUIRES@/${req_bytes}/g" \
-            src/$part/META.in >src/$part/META
-    fi
-done
+tmp="$(mktemp)"
+find src ! -name "$(printf "*\n*")" -name 'META.in' -type f > "$tmp"
```
Comment (gerdstolpmann) [inline configure:689]: the part `-name "$(printf ...)"` looks questionable. What is the intended meaning?

File: configure [hunk 2]
```diff
@@ -676,32 +677,33 @@ for lib in $generated_META $lbytes; do
...
+while IFS= read -r part
+do
+    out=$(echo "$part" | sed -e "s/\.in\$//")
```
Comment (gerdstolpmann) [inline configure:692]: `out=${part%.in}` perhaps?

File: configure [hunk 3]
```diff
@@ -728,43 +730,46 @@ if [ $cbytes -gt 0 ]; then
...
+        echo "OCAMLOPT_FLAGS=-I \"${mingw_lib}\""
+    fi
+    echo "OPAQUE=${opaque}"
+    echo "OCAMLOPT_G=${native_debugging_info}"
+    echo "CHECK_BEFORE_INSTALL=${check_before_install}"
+    echo "INSTALLDIR = install -d"
+    echo "# change to INSTALLDIR = mkdir -p   when BSD install is unavavailable"
+    echo "INSTALLFILE = install -c"
+    echo "# change to INSTALLFILE = cp   when BSD install is unavailable"
+} >> Makefile.config
```
Comment (gerdstolpmann) [inline configure:771]: no, if you combine all the outputs it is `> Makefile.config`

File: configure [hunk 4]
```diff
@@ -728,43 +730,46 @@ if [ $cbytes -gt 0 ]; then
...
-fi
-echo "OPAQUE=${opaque}" >>Makefile.config
-echo "OCAMLOPT_G=${native_debugging_info}" >>Makefile.config
-echo "CHECK_BEFORE_INSTALL=${check_before_install}" >>Makefile.config
-echo "INSTALLDIR = install -d" >>Makefile.config
-echo "# change to INSTALLDIR = mkdir -p   when BSD install is unavavailable" >>Makefile.config
-echo "INSTALLFILE = install -c" >>Makefile.config
-echo "# change to INSTALLFILE = cp   when BSD install is unavailable" >>Makefile.config
+{
+    echo "# Makefile.config written by configure" >Makefile.config
```
Comment (gerdstolpmann) [inline configure:734]: double redirection

File: configure (hunk 1 above)
Comment (Leonidas-from-XIV) [reply configure:689]: it is `! -name $(printf ...)` to skip over names that have newlines. The issue is [described as SC2044](https://www.shellcheck.net/wiki/SC2044) and a better readable solution is possible but only by using bashisms. Not that we expect files with newlines here…

The other possible portable POSIX solution is to use `find`s `-exec`, but I am not sure how readable a longer `exec` command is.

File: configure (hunk 2 above)
Comment (Leonidas-from-XIV) [reply configure]: Indeed this is better. I had thought this was a bash-ishm but I tested with `dash` and `shellcheck` and it works!

File: configure (hunk 4 above)
Comment (Leonidas-from-XIV) [reply configure]: I had this to make sure not to append to an existing `Makefile.config` but as you point out, redirecting the whole thing via `>` makes more sense. Somehow missed the forest for the trees.

File: configure (hunk 1 above)
Comment (Leonidas-from-XIV) [reply configure]: I've ended up rewriting this using `-exec` (as described in the wiki above) as this gets rid of the temporary files and with the `${part%.in}` variable substitution form it is halfway readable.

File: configure [hunk 5]
```diff
@@ -196,14 +196,14 @@ for tool in sed ocaml ocamlc uname rm make cat dirname basename; do
     fi
 done
 
-lib_suffix=`ocamlc -config 2>/dev/null|tr -d '\015'|sed -n -e 's/^ext_lib: //p'`
+lib_suffix=$(ocamlc -config 2>/dev/null | tr -d '\015' | sed -n -e 's/^ext_lib: //p')
```
Comment (dra27) [inline configure:199]: We could POSIX-ly clearer here and have:
```suggestion
lib_suffix=$(ocamlc -config 2>/dev/null | tr -d '\r' | sed -n -e 's/^ext_lib: //p')
```
(and in other places)

File: configure [hunk 6]
```diff
@@ -584,7 +584,7 @@ fi
 have_natdynlink=0
 natdynlink=""
 camlp4_dynlink=""
-if [ -f "${ocaml_core_stdlib}/${dynlink_subdir}/dynlink.cmxa" ]; then
+if [ -f "${ocaml_core_stdlib}/dynlink.cmxa" ]; then
```
Comment (dra27) [inline configure:584]: While it has some obvious benefits, this is the problem I have with shellcheck - it becomes too easy to believe.

This regresses #49 - the variable is set here:

https://github.com/ocaml/ocamlfind/blob/c14912adb4c32d1179806b649215ffe49234c37b/configure#L528

File: configure [hunk 7]
```diff
@@ -676,32 +677,29 @@ for lib in $generated_META $lbytes; do
         -e "s|%%interfaces%%|${if}|g" \
         -e "s|%%findlib_version%%|${version}|g" \
         -e "s|%%natdynlink%%|${natdynlink}|g" \
-        -e "s|%%dynlink_dir%%|${dynlink_dir}|g" \
-        -e "s|%%unix_dir%%|${unix_dir}|g" \
-        -e "s|%%str_dir%%|${str_dir}|g" \
-        site-lib-src/$lib/META.in > site-lib-src/$lib/META
+        -e "s|%%dynlink_dir%%||g" \
+        -e "s|%%unix_dir%%||g" \
+        -e "s|%%str_dir%%||g" \
```
Comment (dra27) [inline configure:682]: Similarly, this breaks the script (cf. a 4.14 compilation)

**Review by dra27 — CHANGES_REQUESTED** [review]
I thought this might be quickly rebasable, but on closer inspection it breaks the script.

There are several bits where the indentation has become odd, possibly because your editor objects to tabs? It might be worth normalising the whitespace as an initial commit if it gets redone.

**Comment by zapashcanon** [comment]
FYI, I tried to change the quoting stuff (*i.e.* `` ` ` `` vs `$()`) in the OCaml codebase and it supposedly break some systems in `/bin/sh` scripts, see [this comment](https://github.com/ocaml/ocaml/pull/1690#issuecomment-377928757).

**Comment by dra27** [comment]
6 years later, the quoting concerns for backticks are somewhat less of a concern - I think the intersection of "has a C11 compiler and can therefore build OCaml 5.x" and "has a non-Posix sh" is fortunately ∅ 😊

File: configure (hunk 5 above)
Comment (Leonidas-from-XIV) [reply configure]: Thanks, I applied this to all occurences of `\015`!

File: configure (hunk 7 above)
Comment (Leonidas-from-XIV) [reply configure]: Good point, I've fixed it in the same way as above, with the `:?` guard. This makes shellcheck happy and I also tried to configure it on 4.14 - it worked there as well.

**Comment by Leonidas-from-XIV** [comment]
I've rebased it on top of the current `master` and applied @dra27's very helpful comments.

As to the `$()` thing, I've left it in. While the OCaml compiler can probably happily depend on `$()`, ocamlfind does support 3.08 so might encounter these systems. On the other hand, I would not even know where to find a system with a non-POSIX `sh` so I think requiring a shell that's newer than the last few decades is probably fine. We don't have a good way to test whether the rest of the script works anyway on such platforms.

**Pull request merged**